internal/eventstream/client.go

```diff
@@ -0,0 +1,319 @@
+
+		eventTags = map[string]string{
+			"host": host,
+		}
+	}
+
+	hostGroups, err := client.fetchHostGroups(host)
+	if err != nil {
+		return nil, err
+	}
+	for _, hostGroup := range hostGroups {
+		eventExtraTags["hostgroup/"+hostGroup] = ""
+	}
+
+	return &event.Event{
+		Time:      time.Now(),
+		SourceId:  client.IcingaNotificationsEventSourceId,
+		Name:      eventName,
+		URL:       client.IcingaWebRoot + eventUrlSuffix,
+		Tags:      eventTags,
+		ExtraTags: eventExtraTags,
+	}, nil
+}
+
+// buildHostServiceEvent constructs an event.Event based on a CheckResult, a Host or Service state, a Host name and an
+// optional Service name if the Event should represent a Service object.
+func (client *Client) buildHostServiceEvent(result CheckResult, state int, host, service string) (*event.Event, error) {
+	var eventSeverity event.Severity
+
+	if service != "" {
+		switch state {
+		case 0:
+			eventSeverity = event.SeverityOK
+		case 1:
+			eventSeverity = event.SeverityWarning
+		case 2:
+			eventSeverity = event.SeverityCrit
+		default:
+			eventSeverity = event.SeverityErr
+		}
+	} else {
+		switch state {
+		case 0:
+			eventSeverity = event.SeverityOK
+		case 1:
+			eventSeverity = event.SeverityCrit
+		default:
+			eventSeverity = event.SeverityErr
+		}
+	}
+
+	ev, err := client.buildCommonEvent(host, service)
+	if err != nil {
+		return nil, err
+	}
+
+	ev.Type = event.TypeState
+	ev.Severity = eventSeverity
+	ev.Message = result.Output
+
+	return ev, nil
+}
+
+// buildAcknowledgementEvent from the given fields.
+func (client *Client) buildAcknowledgementEvent(host, service, author, comment string) (*event.Event, error) {
+	ev, err := client.buildCommonEvent(host, service)
+	if err != nil {
+		return nil, err
+	}
+
+	ev.Type = event.TypeAcknowledgement
+	ev.Username = author
+	ev.Message = comment
+
+	return ev, nil
+}
+
+// eventDispatcher receives generated event.Events to be either buffered or directly delivered to the CallbackFn.
+//
+// When the Client is in the reconnection phase, indicated in the enterReconnectionPhase method, than all received Events
+// from the eventDispatch channel will be buffered until the replayTrigger fires.
+func (client *Client) eventDispatcher() {
+	var reconnectionBuffer []*event.Event
+
+	for {
+		select {
+		case <-client.Ctx.Done():
+			client.Logger.Warnw("Closing event dispatcher as context is done", zap.Error(client.Ctx.Err()))
+			return
+
+		case <-client.replayTrigger:
+			for _, ev := range reconnectionBuffer {
+				client.CallbackFn(ev)
+			}
+			client.Logger.Debugf("Replayed %d events during reconnection phase", len(reconnectionBuffer))
+			client.replayPhase.Store(false)
+			reconnectionBuffer = []*event.Event{}
+			client.Logger.Info("Finished reconnection phase and returning normal operation")
+
+		case ev := <-client.eventDispatch:
+			if client.replayPhase.Load() {
+				reconnectionBuffer = append(reconnectionBuffer, ev)
+			} else {
+				client.CallbackFn(ev)
+			}
+		}
+	}
+}
```

Comment:
If I didn't overlook something, you seem to be buffering the events from both steps (1) and (2) mentioned in https://github.com/Icinga/icinga-notifications/pull/112#issuecomment-1780814041 here.

My idea behind the buffering was to ensure a consistent order of the events we are processing. So a simplified version of what I would want to do would be:
1. Query all state we need to know form the objects API and establish this a the current state of all objects/incidents here by emitting corresponding events.
2. After that's done, process further updates from the event stream.

Now if that request for (2) was started only after (1) finished, events may be lost in between, therefore, it has to be started before (1), so that we know we receive all events and buffer them for future use (but only those from the event stream). However, now we may also receive some events that are older than the data read in (1), but these can be removed based on the timestamps on a per-object basis.